### ISSUE TITLE
New CA2007 fails on 'await using' of an IAsyncDisposable

### ISSUE THREAD

**Issue by DaRosenberg** [dotnet/roslyn-analyzers#5712]
### Analyzer

**Diagnostic ID**: [CA2007](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2007): `Consider calling ConfigureAwait on the awaited task`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 6 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.100](https://dotnet.microsoft.com/download/dotnet/6.0)

### Describe the bug

CA2007 in .NET 6 SDK has started giving what I feel are a bunch of false positives on `await using` constructs of types that implement `IAsyncDisposable`.

### Steps To Reproduce

1. In C#, write an `async using` statement on any `IAsyncDisposable`

### Expected behavior

Analyser should not flag this case because I don't think there's any way to correct.

### Actual behavior

Analyser emits a warning, and suggests a fix which makes the code not compile:

<img width="2443" alt="image" src="https://user-images.githubusercontent.com/3889090/141119002-2ab84bdd-6964-4657-95e3-cb161642fb4b.png">

With this fix, the type of the variable changes to `ConfiguredAsyncDisposable` which leads to compile errors in subsequent code that tries to use this variable for what it really is.

**Comment by DaRosenberg** [comment]
Researching this a bit more I can see that the analyser warning it self seems to be very much by design: #4888 

But still, the code fix should not introduce compiler errors. The fix should produce the recommended code pattern instead:

```cs
var storage = await persistService.GetStorageAsync(solution, cancellationToken).ConfigureAwait(false);
await using var _ = storage.ConfigureAwait(false);
```

Updating issue accordingly.

**Comment by manfred-brands** [comment]
Unfortunately, moving to a two part statement raises CA2000.

![image](https://user-images.githubusercontent.com/8655218/142322233-a46dbe80-0277-4aaa-b201-1866e84248b5.png)

**Comment by DaRosenberg** [comment]
@manfred-brands Yes, neither is ideal. CA2000 gives a false positive in this situation IMO, it should probably be extended to detect that this code actually does dispose of the object. Another warning is better than broken code, though...

**Comment by daviddunson** [comment]
After upgrading an existing project from VS 2019 & .NET 5 to VS 2022 and .NET 6, I get a lot of false positives on CA2007.

```
await using var command = connection.CreateCommand(); // <= CA2007
await using var reader = await command.ExecuteReaderAsync(CommandBehavior.SingleResult, cancellationToken).ConfigureAwait(false); // <= CA2007
```

I'm using System.Data.SQLite.SQLiteConnection.

**Comment by amramz** [comment]
Any updates / ETA on this one?
Thanks

**Comment by virzak** [comment]
Just ran into the same issue, and it's been exactly a year since last comment, so nudge.

**Comment by NinjaCross** [comment]
Any news on this ?
Another year passed, and the problem is still present :(

**Comment by robsyn-nordsec** [comment]
4 more months have passed, can we have an update on this please? :)

**Comment by mavasani** [comment]
@carlossanlop @buyaa-n for triage

Folks - very sorry for the delay here. However, note that this repo is heavily community driven. We would definitely take a community PR for this fix. I'll let Carlos and Buyaa decide if the feedback here warrants a dev member from dotnet org to work on this sooner.

**Comment by mavasani** [comment]
Also tagging @sharwell as the author of #4896

**Comment by sharwell** [comment]
> Unfortunately, moving to a two part statement raises CA2000: 'Call Dispose on object before all references are out of scope' That leaves us in a loose-loose situation.
> 
> ![image](https://user-images.githubusercontent.com/8655218/142322233-a46dbe80-0277-4aaa-b201-1866e84248b5.png)

This would be a false positive for CA2000. I generally do not recommend enabling CA2000, as there is no established coding pattern which allows it to work without notable false positives or false negatives.

**Comment by daviddunson** [comment]
> This would be a false positive for CA2000. I generally do not recommend enabling CA2000, as there is no established coding pattern which allows it to work without notable false positives or false negatives.

Not an option. My company requires 100% compliance. Cannot disable CA2000 for other use cases. This only appears when using a class with a base class that implements IDisposable. i.e Anything that inherits from Stream. I've gotten used to isolating these calls in to a very small method and suppressing that instance of the CA2000, so that CA2000 can report legitimate warnings.

**Comment by sharwell** [comment]
> After upgrading an existing project from VS 2019 & .NET 5 to VS 2022 and .NET 6, I get a lot of false positives on CA2007.
> 
> ```
> await using var command = connection.CreateCommand(); // <= CA2007
> await using var reader = await command.ExecuteReaderAsync(CommandBehavior.SingleResult, cancellationToken).ConfigureAwait(false); // <= CA2007
> ```
> 
> I'm using System.Data.SQLite.SQLiteConnection.

This should be filed as a separate issue. The current issue relates only to a detail of the code fix which applies a change after CA2007 is reported.
